clients/src/test/java/org/apache/kafka/clients/consumer/internals/MembershipManagerImplTest.java

```diff
@@ -17,253 +17,1059 @@
 
+    /**
+     * This test should be the case when an assignment is sent to the member, and it cannot find
+     * it in metadata (permanently, ex. topic deleted). The member will keep the assignment as
+     * waiting for metadata, but the following assignment received from the broker will not
+     * contain the deleted topic. The member will discard the assignment that was pending and
+     * proceed with the reconciliation of the new assignment.
+     */
     @Test
-    public void testAssignmentUpdatedAsReceivedAndProcessed() {
-        MembershipManagerImpl membershipManager = new MembershipManagerImpl(GROUP_ID, logContext);
-        ConsumerGroupHeartbeatResponseData.Assignment newAssignment = createAssignment();
-        ConsumerGroupHeartbeatResponse heartbeatResponse =
-                createConsumerGroupHeartbeatResponse(newAssignment);
-        membershipManager.updateState(heartbeatResponse.data());
+    public void testNewAssignmentReplacesPreviousOneWaitingOnMetadata() {
+        MembershipManagerImpl membershipManager = mockJoinAndReceiveAssignment(false);
+        assertEquals(MemberState.RECONCILING, membershipManager.state());
+
+        // When the ack is sent the member should go back to RECONCILING
+        membershipManager.onHeartbeatRequestSent();
+        assertEquals(MemberState.RECONCILING, membershipManager.state());
+        assertTrue(membershipManager.topicsWaitingForMetadata().size() > 0);
+
+        // New target assignment received while there is another one waiting to be resolved
+        // and reconciled. This assignment does not include the previous one that is waiting
+        // for metadata, so the member will discard the topics that were waiting for metadata, and
+        // reconcile the new assignment.
+        Uuid topicId = Uuid.randomUuid();
+        String topicName = "topic1";
+        when(metadata.topicNames()).thenReturn(Collections.singletonMap(topicId, topicName));
+        receiveAssignment(topicId, Collections.singletonList(0), membershipManager);
+        Set<TopicPartition> expectedAssignment = Collections.singleton(new TopicPartition(topicName, 0));
+        assertEquals(MemberState.ACKNOWLEDGING, membershipManager.state());
+        verify(subscriptionState).assignFromSubscribed(expectedAssignment);
+
+        // When ack for the reconciled assignment is sent, member should go back to STABLE
+        // because the first assignment that was not resolved should have been discarded
+        membershipManager.onHeartbeatRequestSent();
+        assertEquals(MemberState.STABLE, membershipManager.state());
+        assertTrue(membershipManager.topicsWaitingForMetadata().isEmpty());
+    }
+
+    /**
+     * This test should be the case when an assignment is sent to the member, and it cannot find
+     * it in metadata (temporarily). The broker will continue to send the assignment to the
+     * member. The member should keep it was waiting for metadata and continue to request updates.
```

Comment:
`The broker will continue to send the assignment to the member.` This is not entirely true. The broker may not send anything.

Reply:
Agree, rephrased it. This test is just for the case where the broker does keep sending (and the test above for when it does not)